src/index.js

```diff
@@ -27,95 +27,80 @@ export function h(type, props) {
   var lock
   var root = (container = container || document.body).children[0]
   var node = vnode(root, [].map)
-  var lifecycle = []
-  var appState = props.state || {}
-  var appActions = {}
+  var stack = []
 
-  repaint(init(appState, appActions, props.actions, []))
+  repaint(init(model, []))
 
-  return appActions
+  return model
 
   function vnode(element, map) {
     return (
-      element &&
-      h(
-        element.tagName.toLowerCase(),
-        {},
-        map.call(element.childNodes, function(element) {
-          return element.nodeType === 3
+      element && {
+        type: element.tagName.toLowerCase(),
+        props: {},
+        children: map.call(element.childNodes, function(element) {
+          return 3 === element.nodeType
             ? element.nodeValue
             : vnode(element, map)
         })
-      )
+      }
     )
   }
 
-  function set(to, from) {
-    for (var i in from) {
-      to[i] = from[i]
+  function set(target, source) {
+    for (var i in source) {
+      target[i] = source[i]
     }
-    return to
+    return target
   }
 
-  function merge(to, from) {
-    return set(set({}, to), from)
+  function merge(target, source) {
+    return set(set({}, target), source)
   }
 
-  function setDeep(path, value, from) {
-    var to = {}
+  function setDeep(path, value, source) {
+    var target = {}
     return 0 === path.length
       ? value
-      : ((to[path[0]] =
+      : ((target[path[0]] =
           1 < path.length
-            ? setDeep(path.slice(1), value, from[path[0]])
+            ? setDeep(path.slice(1), value, source[path[0]])
             : value),
-        merge(from, to))
+        merge(source, target))
   }
 
-  function get(path, from) {
+  function get(path, source) {
     for (var i = 0; i < path.length; i++) {
-      from = from[path[i]]
+      source = source[path[i]]
     }
-    return from
-  }
-
-  function isFunction(any) {
-    return "function" === typeof any
+    return source
   }
 
-  function init(state, actions, from, path) {
-    for (var key in from) {
-      isFunction(from[key])
+  function init(source, path) {
+    for (var key in source) {
+      typeof source[key] === "function"
         ? (function(key, action) {
-            actions[key] = function(data) {
-              state = get(path, appState)
-
-              if (
-                isFunction((data = action(data))) &&
-                isFunction((data = data(state)))
-              ) {
-                data = data(actions)
+            source[key] = function(data) {
+              source = get(path, model)
+
+              if (typeof (data = action(data)) === "function") {
+                data = data(source)
               }
 
-              if (data && data !== state && !data.then) {
-                repaint(
-                  (appState = setDeep(path, merge(state, data), appState))
-                )
+              if (data && data !== source && !data.then) {
```

Comment:
I just thought of something here...
If you also check `&& !(data.type && data.props && data.children)` you can allow actions to return VNodes, i.e. wired components for cheap.

Reply:
@Mytrill Let me see.

Reply:
This felt awfully fitting with respect to the API changes proposed here. Awesome that it could be done with such a small change to core.

Reply:
@lukejacksonn @Mytrill It works, but also complicates the API. I need to think more about it. 😄